### PR TITLE
Unit test for progressRevertStake

### PR DESCRIPTION
**Pull request by deepesh-kn**
⚠️ Blocked by #578 
This PR contains unit test of EIP20Gateway.progressRevertStake()
Fixes: #426

### CODE REVIEW

File: contracts/gateway/EIP20Gateway.sol [hunk 1]
```diff
@@ -572,18 +572,18 @@ contract EIP20Gateway is GatewayBase {
...
-            "RLP parent nodes must not be zero"
+            "RLP parent nodes must not be zero."
         );
 
         // Get the message object
         MessageBus.Message storage message = messages[_messageHash];
         require(
             message.intentHash != bytes32(0),
-            "StakeIntentHash must not be zero"
+            "StakeIntentHash must not be zero."
```
Comment (sarvesh-ost) [inline contracts/gateway/EIP20Gateway.sol]: Please add a period to L586.

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+const ZeroBytes = Utils.ZERO_BYTES32;
+const MessageStatusEnum = messageBus.MessageStatusEnum;
+
+contract('EIP20Gateway.progressRevertStake()', function (accounts) {
+
+  let gateway, mockToken, baseToken, revertStakeParams, bountyAmount, gatewayLib;
+
+  let setup = async function(revertStakeParams){
+
+    let stakeIntentHash = await gatewayLib.hashStakeIntent(
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: Please use the available helper for this.

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+      rlpParentNodes: revokedProofData.co_gateway.confirm_revert_stake_intent.proof_data.storageProof[0].serializedProof,
+      blockHeight: new BN(revokedProofData.co_gateway.confirm_revert_stake_intent.proof_data.block_number,16),
+      gatewayAddress: revokedProofData.contracts.gateway,
+      storageRoot: revokedProofData.co_gateway.confirm_revert_stake_intent.proof_data.storageHash,
+    };
+
+    await setup(revertStakeParams);
+
+    await mockToken.transfer(gateway.address, revertStakeParams.amount, { from: accounts[0] });
+    await baseToken.transfer(gateway.address, bountyAmount.muln(2.5), { from: accounts[0] });
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: Add documentation for 2.5 param.

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+  });
+
+  it('should fail when message hash is zero', async function () {
+
+    await gateway.setStorageRoot(revertStakeParams.blockHeight, revertStakeParams.storageRoot);
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: is this required?

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+        revertStakeParams.messageHash,
+        blockHeight,
+        revokedProofData.co_gateway.confirm_stake_intent.proof_data.storageProof[0].serializedProof,
+      ),
+      'Merkle proof verification failed.',
+    );
+
+  });
+
+  it('should fail when storage proof is invalid', async function () {
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: I am not sure if we need this test. This seems like testing of MPP library.

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+        revertStakeParams.messageHash,
+        revertStakeParams.blockHeight,
+        revertStakeParams.rlpParentNodes,
+      ),
+      'Message on source must be DeclaredRevocation.',
+    );
+
+  });
+
+  it('should fail when message status of inbox at target is declared', async function () {
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: This and other tests around status of inbox  seems duplicate of `should fail when storage proof is incorrect`

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+    assert.strictEqual(
+      result.staker_,
+      revertStakeParams.staker,
+      `Staker address must be equal to ${revertStakeParams.staker}.`,
+    );
+
+    assert.strictEqual(
+      result.amount_.eq(revertStakeParams.amount),
+      true,
+      `Stake amount must be equal to ${revertStakeParams.amount.toString(10)}.`,
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: Add both values of assert in the reason. Please check other asserts as-well

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+      burnerFinalTokenBalance.eq(burnerInitialTokenBalance.add(bountyAmount.muln(2.5))),
+      true,
+      `Burner balance must increase by ${bountyAmount.muln(2.5).toString(10)}.`,
+    );
+
+  });
+
+  it('should transfer stake amount to staker ', async function () {
+
+    let burnerAddress = await gateway.burner.call();
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: Why burner is important in this context?

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+    );
+
+    assert.strictEqual(
+      burnerFinalTokenBalance.eq(burnerInitialTokenBalance),
+      true,
+      `Burner balance must not change.`,
+    );
+
+  });
+
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]:
```suggestion
```

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+
+  it('should fail when message status of inbox at target is progressed', async function () {
```
Comment (sarvesh-ost) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: Please group negative and positive tests.

File: contracts/gateway/EIP20Gateway.sol (hunk 1 above)
Comment (deepesh-kn) [reply contracts/gateway/EIP20Gateway.sol]: Is it not already added ?

File: test/gateway/eip20_gateway/progress_revert_stake.js (hunk 5 above)
Comment (deepesh-kn) [reply test/gateway/eip20_gateway/progress_revert_stake.js]: It is not a pure unit test, its an integration of Gateway+GatewayLib+MessageBus+MPP as a unit.

File: test/gateway/eip20_gateway/progress_revert_stake.js (hunk 8 above)
Comment (deepesh-kn) [reply test/gateway/eip20_gateway/progress_revert_stake.js]: As there is a transfer of base token to burner address in revert progress. 
So it makes sense to assert the EIP20token balance to make sure that there is no transfer of EIP20 tokens to burner address.

File: contracts/gateway/EIP20Gateway.sol (hunk 1 above)
Comment (schemar) [reply contracts/gateway/EIP20Gateway.sol]: image
I can see it 🤔

File: test/gateway/eip20_gateway/progress_revert_stake.js (hunk 5 above)
Comment (schemar) [reply test/gateway/eip20_gateway/progress_revert_stake.js]: Integration of multiple units is generally not a unit, but an integration 🤔

File: test/gateway/eip20_gateway/progress_revert_stake.js [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+
+    assert.equal(
+      tx.receipt.status,
+      1,
+      "Receipt status is unsuccessful",
+    );
+
+  });
+
+  it('should emit `StakeReverted` event ', async function () {
```
Comment (schemar) [inline test/gateway/eip20_gateway/progress_revert_stake.js]: Some have extra spaces again 🧐
```suggestion
  it('should emit `StakeReverted` event', async function () {
```

File: test/gateway/eip20_gateway/progress_revert_stake.js (hunk 5 above)
Comment (deepesh-kn) [reply test/gateway/eip20_gateway/progress_revert_stake.js]: Yes, I understand this but we decided not to replicate code with `MockMessageBus` approach,  so all the progress with the proof tests is integration w.r.t `messageBus`, `gatewayLib` and `MPP`.

In such a scenario, I am not sure where to draw a hard line for what to test and not to. So I add the tests considering all as a unit. 

I am in favor to keep such tests where ever we have integration. But if the majority of us decides that it is not needed then we can remove it.

**Pull request merged**